// tiled-tmx-parser/dist/classes/TiledMap.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TiledMap = void 0;
const ObjectLayer_1 = require("./ObjectLayer");
class TiledMap {
    constructor(urlOrPath) {
        this.urlOrPath = urlOrPath;
        this.layers = [];
    }
    getTileSetByGid(gid) {
        return this.tileSets.find((ts) => {
            const minGid = ts.firstGid;
            const maxGid = ts.firstGid + ts.tiles.length - 1;
            return gid >= minGid && gid <= maxGid;
        });
    }
    getTileById(id) {
        const tileSet = this.getTileSetByGid(id);
        if (tileSet) {
            return tileSet.tiles.find(x => x.id === id);
        }
        return;
    }
    getObjectById(id) {
        let object;
        for (let i = 0; i < this.layers.length; i++) {
            const layer = this.layers[i];
            if (layer instanceof ObjectLayer_1.ObjectLayer) {
                object = layer.objects.find(x => x.id === id);
                if (object) {
                    return object;
                }
            }
        }
        return object;
    }
}
exports.TiledMap = TiledMap;
